File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/LocalDateTimeConverter.java
package com.thewizrd.shared_resources.utils;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

public class LocalDateTimeConverter extends TypeAdapter<LocalDateTime> {
    private static final DateTimeFormatter formatter = DateTimeFormatter.ISO_INSTANT;

    @Override
    public LocalDateTime read(JsonReader in) throws IOException {
        return LocalDateTime.ofInstant(Instant.from(formatter.parse(in.nextString())), ZoneOffset.UTC);
    }

    @Override
    public void write(JsonWriter out, LocalDateTime value) throws IOException {
        out.value(value.toInstant(ZoneOffset.UTC).toString());
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/json/LocalDateTimeConverter.kt
package com.thewizrd.shared_resources.json

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

class LocalDateTimeConverter : JsonAdapter<LocalDateTime>() {
    companion object {
        private val formatter = DateTimeFormatter.ISO_INSTANT
    }

    @Synchronized
    override fun fromJson(reader: JsonReader): LocalDateTime? {
        if (reader.peek() == JsonReader.Token.NULL) {
            return reader.nextNull()
        }

        return LocalDateTime.ofInstant(
            Instant.from(formatter.parse(reader.nextString())), ZoneOffset.UTC
        )
    }

    @Synchronized
    override fun toJson(writer: JsonWriter, value: LocalDateTime?) {
        if (value == null) {
            writer.nullValue()
        } else {
            writer.value(value.toInstant(ZoneOffset.UTC).toString())
        }
    }
}